R/validator.R

```diff
@@ -0,0 +1,85 @@
+##
+## Author: Chris Bielow
+##
+
+## Functions to validate an mzQC file or in-memory object or an mzQC JSON string
+
+
+#'
+#' Get a validator for mzQC
+#'
+#' @import jsonvalidate
+#'
+getValidator = function()
+{
+  schema = system.file("schema/mzQC_schema.json", package = "rmzqc", mustWork = TRUE)
+  v = jsonvalidate::json_validator(schema, engine = "ajv") ## we need schema draft-07
+  return(v)
+}
+
+#'
+#' Validates an mzQC document which is present as a file.
+#'
+#' The returned TRUE/FALSE has additional attributes in case of errors.
+#' Use attributes(result) to access them.
+#'
+#' @param filepath A path to a file (e.g. "c:/my.mzQC", or "test.mzQC")
+#' @param verbose Show extra information if validation fails
+#' @return TRUE/FALSE if validation was successful/failed
+#'
+#' @export
+#'
+validateFromFile = function(filepath, verbose = TRUE)
+{
+  data = readLines(filepath)
+  validateFromString(data, verbose)
+}
+
+
+#'
+#' Validates an mzQC document which is already in memory as JSON string.
+#' e.g. the string "{ mzQC : {}}"
+#'
+#' If the string object passed into this function contains multiple elements (length > 1).
+#' then they will be concatenated using '\\n' before validation.
+#'
+#' The returned TRUE/FALSE has additional attributes in case of errors.
+#' Use attributes(result) to access them.
+#'
+#' @param JSON_string A string which contains JSON (multiple lines allowed)
+#' @param verbose Show extra information if validation fails
+#' @return TRUE/FALSE if validation was successful/failed
+#'
+#' @export
+#'
+validateFromString = function(JSON_string, verbose = TRUE)
+{
+  v = getValidator()
+  if (length(JSON_string) > 1) JSON_string = paste0(JSON_string, collapse = "\n") ## we need a single line for the validator
+  res = v(JSON_string, verbose = verbose)
+  return(res)
+}
+
+
+#'
+#' Validates an mzQC document which is already in memory as mzQC root object, as obtained by, e.g. readMZQC().
+#'
+#' This method is less performant than validateFromString,
+#' because it needs to convert the R object to
+#' a JSON string first.
+#'
+#' The returned TRUE/FALSE has additional attributes in case of errors.
+#' Use attributes(result) to access them.
+#'
+#' @param mzqc_root An mzQC root object
+#' @param verbose Show extra information if validation fails
+#' @return TRUE/FALSE if validation was successful/failed
+#'
+#' @export
+#'
+validateFromObj = function(mzqc_root, verbose = TRUE)
+{
+  if (!("MzQCmzQC" %in% class(mzqc_root))) stop("Argument is not a class of 'MzQCmzQC'.")
```

Comment:
Sounds reasonable. Maybe an additional check if toJSON was successful (i.e. valid json)?

Reply:
This is done 'automatically' during `::toJSON`. Users will see an error message and a backtrace.

Reply:
However, this reminds me that one could check if the JSON itself when reading from a file/string is valid.
Done.

Before (not very informative)
```
res = validateFromString(strings)
Error in context_eval(join(src), private$context, serialize, await) : 
SyntaxError: Invalid or unexpected token
```

After:
```
>   res = validateFromString(strings)
Warning in validateFromString(strings) :
  925lexical error: invalid character inside string.
           software for QC report genera             }           ]    
                     (right here) ------^

Error in validateFromString(strings) : 
  The JSON string is not valid JSON.
  ```

Reply:
:+1